File: 2021/05/06/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        
        def sortToBST(begin, end):
            if begin == end:
                return None
            
            if begin.next == end:
                return TreeNode(begin.val)
            
            fast, slow = begin, begin
            while fast != end and fast.next != end:
                fast = fast.next.next
                slow = slow.next
            mid = slow
            
            return TreeNode(mid.val, sortToBST(begin, mid), sortToBST(mid.next, end))
        
        return sortToBST(head, None)

File: 2021/05/06/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedListToBST(head *ListNode) *TreeNode {
	return sortToBST(head, nil)
}

func sortToBST(begin *ListNode, end *ListNode) *TreeNode {
	if begin == end {
		return nil
	}
	if begin.Next == end {
		return &TreeNode{Val: begin.Val}
	}
	fast, slow := begin, begin
	for fast != end && fast.Next != end {
		fast = fast.Next.Next
		slow = slow.Next
	}
	mid := slow
	return &TreeNode{
		Val:   mid.Val,
		Left:  sortToBST(begin, mid),
		Right: sortToBST(mid.Next, end),
	}
}
